Improve content for beginner SQL content (#1405)
enkidevs/curriculum: sql/dql/basic-queries/where-3.md

---
## Content

### `LIKE`, `IN`, `BETWEEN`

SQL provides a bunch of special operators to help you write complex querying conditions. More specifically, in this insight, we will take a look at:

> - the `LIKE` operator which is used to define a pattern that the given attribute should conform to

Reviewer: I think putting a comma between "operator" and "which" for all of these would read better ([and is grammatically correct](https://www.grammarly.com/blog/comma-before-which/)).

> We'll be using:

> A [pause] which does X,
> B [pause] which does Y,
> C [pause] which does Z

| Charmeleon | 405   | 58 | 64     | 58      |
| Charizard  | 534   | 78 | 84     | 78      |
| Squirtle   | 314   | 44 | 48     | 65      |
| Wartortle  | 405   | 59 | 63     | 80      |
| Blastoise  | 530   | 79 | 83     | 100     |
| Caterpie   | 195   | 45 | 30     | 35      |
| Metapod    | 205   | 50 | 20     | 55      |

The `IN` condition allows you to easily test if a value inside a column matches any one specific item in a list of values.

```sql
SELECT *
FROM pokemon
WHERE name IN ('Charmander',
  'Squirtle', 'Bulbasaur');
```

In this case, we will only get the rows of the three starter Pokémon: Charmander, Squirtle and Bulbasaur.

| name       | total | hp | attack | defense |
|------------|-------|----|--------|---------|
| Bulbasaur  | 318   | 45 | 49     | 49      |
| Charmander | 309   | 39 | 52     | 43      |
| Squirtle   | 314   | 44 | 48     | 65      |

The `LIKE` statement accepts a pattern string, where wildcard characters are used to denote an acceptable pattern.[1]

 - `%` is a **wildcard**. This means it counts for a string of any characters of any length, including nothing.
 - `_` represents only one character.

> Therefore, `%le` denotes any string that ends in 'le'. `_le` denotes any string with only one character before 'le'. They can be chained together into complex patterns like `___taco%` which denotes any string containing 'taco' starting on the 4th character and with any number of characters after.

Reviewer 2: ```suggestion
Therefore, `%le` denotes any string that ends in 'le'. `_le` denotes any string with only one character before 'le'. They can be chained together into complex patterns like `___taco%` which denotes any string containing the substring 'taco' at its 4th position, and with any number of characters after.
```

```sql
SELECT *
FROM pokemon
WHERE name LIKE '%le';
```

> The result of this statement is the two rows of the Pokémon whose names end in '-le': Squirtle and Wartortle.

Reviewer 2: ```suggestion
The result of this statement is the two rows of the Pokémon whose names end in 'le': Squirtle and Wartortle.
```

| name      | total | hp | attack | defense |
|-----------|-------|----|--------|---------|
| Squirtle  | 314   | 44 | 48     | 65      |
| Wartortle | 405   | 59 | 63     | 80      |

> Instead of chaining two comparison operators in your `WHERE` clause you can use the `BETWEEN` operator to define a range of acceptable values, like so:

Reviewer 2: > Instead of chaining two comparison operators in your `WHERE` clause

I don't think you should start with this as it's not obvious how BETWEEN is equivalent to that. maybe 

add a short paragraph as a footnote:

"This can be seen as a shortcut for chaining two comparison operators with the `AND` clause"
-> show the equivalent code

Author: added the following footnote

This can be seen as a shortcut for chaining two comparison operators with the `AND` clause:
```sql
SELECT * FROM table_name
WHERE col BETWEEN 1 AND 10;
-- is the same as
SELECT * FROM table_name
WHERE col >= 1 AND
      col <= 10;
```

> ### `LIKE`, `IN`

Reviewer 2: some learn more links with better explanations would be helpful here
e.g. there should be one about the in-depth LIKE operators and all its options

> In a `WHERE` clause, we can also use the keywords `LIKE` and `IN`. `LIKE` allows us to define a pattern that the given attribute should conform to, such as the length or the first letter. `IN` allows us to define a list of values of which the attribute must be one. 

Reviewer 2: a better introduction for this would be 

> SQL provides a bunch of special operators to help you write complex querying conditions. More specifically, in this insight, we will take a look at:
> - the `LIKE` operator which is used for pattern matching ...
> - the `IN` ...
> - the `BETWEEN` ...

* WHERE
* name

> * IN

Reviewer: These answers seem to be obvious. Maybe add a few misleading candidates?

```md
* WHERE
* name
* IN
* ('Groudon', 'Snorlax', 'Wailord')
* SUCH THAT
* IS ONE OF
* ['Groudon', 'Snorlax', 'Wailord']
```
